[ConstraintSystem] (mostly) Finish up solver type cache implementation

xedin opened the pull request:
Based on all of the great work @rudkx has done, solver should only apply _final_ types
to the expression tree when the solution is deduced. This patch makes it so and fixes
tests (mostly in IDE code completion) along the way.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@rudkx There is one thing left which is going to fix 2 crashers related to explicit casts - we still use `getCastTypeLoc().setType` in CSGen to set opened generic type.

rudkx commented:
@xedin I didn't do any work on `TypeLoc` types yet. That seemed small enough to make it a separate task.

I haven't looked at this for a while, but I believe you also need to clear the type on the Expr when it is cached to ensure that we aren't accidentally reading it in places. That was the real hold up the last time I looked at this.

Can you hold off trying to merge this until I get done with the IUO changes. I suspect I've introduced some new `->getType()` and such that will need to be cleaned up with these changes in place, and I don't want to slow that work down at all.

xedin commented:
@rudkx Sure, there is no rush in merging this, just wanted to get it out since tests are fixed. I has planning to take care of `getCastTypeLoc` before merging anyway.

xedin committed:
[ConstraintSystem] Add TypeLoc caching to constraint system

This is useful for explicit casts and type expressions, where
type loc and expression types might be different, and allows
constraint solver to avoid setting opened types to expressions
which resolves multiple crashes.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
/cc @nathawes @benlangmuir

xedin commented:
/cc @rudkx

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -7881,12 +7881,13 @@ static void noteArchetypeSource(const TypeLoc &loc, ArchetypeType *archetype,
-    
+    auto type = loc.getType();
+    if (!type)
+      type = cs.getType(loc);
```
Why is this directly calling `loc.getType()` and then only if getting `null` calling `cs.getType(loc)`?

rudkx commented on test/Compatibility/tuple_arguments.swift:
```diff
@@ -58,10 +58,10 @@ do {
+  var a = 3
+  var b = 4
+  var c = (3)
+  var d = (a, b)
```
It seems like there is something wrong if this is resulting in this type of diagnostic change.

rudkx commented on test/IDE/complete_constructor.swift:
```diff
@@ -119,7 +119,6 @@ func testExplicitConstructors3P() {
 // EXPLICIT_CONSTRUCTORS_3P: Begin completions
 // EXPLICIT_CONSTRUCTORS_3P-DAG: Decl[Constructor]/CurrNominal: ['(']{#(a): Int#}[')'][#ExplicitConstructors3#]{{; name=.+$}}
 // EXPLICIT_CONSTRUCTORS_3P-DAG: Decl[Constructor]/CurrNominal: ['(']{#a: Int#}, {#b: Float#}[')'][#ExplicitConstructors3#]{{; name=.+$}}
-// EXPLICIT_CONSTRUCTORS_3P-DAG: Decl[FreeFunction]/CurrModule/NotRecommended/TypeRelation[Invalid]: freeFunc()[#Void#]{{; name=.+$}}
```
So I am wondering how the code completion changes (which I haven't looked at closely) are related to the type cache. I would expect no behavioral changes from making the type checker store types in a different place.

rudkx commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2153,6 +2153,8 @@ static void eraseOpenedExistentials(Expr *&expr, ConstraintSystem &CS) {
       });
       CS.setType(expr, type);
+      // Set new type to the expression directly.
+      expr->setType(type);
```
It's non-obvious why this is the right thing to do. This seems like an indication that something that should be reading from the type cache isn't.

rudkx commented on lib/Sema/CSApply.cpp:
```diff
@@ -7481,6 +7481,7 @@ namespace {
         } else {
           // For other closures, type-check the body once we've finished with
           // the expression.
+          cs.setExprTypes(closure);
```
Under what circumstances do we need to do this? We should be type checking these bodies independently. They can refer to captured data, but that data should already have a type, right?

rudkx reviewed (changes requested):
I'd like to have a better understanding of some of the changes and why we see behavioral changes here before we merge anything for this.

rudkx reviewed (approved):
We had a chance to sync up on this and everything looks okay for now. It definitely moves things forward re: fixing some crashers and no longer writing types all over the place. We still need to go back and clear expression types, etc. to finish this up.

xedin committed:
[ConstraintSystem] Don't apply types to expressions until solution is found

Resolves: rdar://problem/36744895

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test
